Content.Shared/Ninja/Components/NinjaGlovesComponent.cs

```diff
@@ -0,0 +1,151 @@
+using Content.Shared.Actions;
+using Content.Shared.Actions.ActionTypes;
+using Content.Shared.DoAfter;
+using Content.Shared.Ninja.Systems;
+using Content.Shared.Tag;
+using Content.Shared.Toggleable;
+using Robust.Shared.Audio;
+using Robust.Shared.GameStates;
+using Robust.Shared.Serialization;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype;
+using Robust.Shared.Utility;
+using System.Threading;
+
+namespace Content.Shared.Ninja.Components;
+
+/// <summary>
+/// Component for toggling glove powers.
+/// Powers being enabled is controlled by User not being null.
+/// </summary>
+[RegisterComponent, NetworkedComponent, AutoGenerateComponentState]
+[Access(typeof(SharedNinjaGlovesSystem))]
+public sealed partial class NinjaGlovesComponent : Component
+{
+    /// <summary>
+    /// Entity of the ninja using these gloves, usually means enabled
+    /// </summary>
+    [ViewVariables, AutoNetworkedField]
+    public EntityUid? User;
+
+    /// <summary>
+    /// The action for toggling ninja gloves abilities
+    /// </summary>
+    [DataField("toggleAction")]
+    public InstantAction ToggleAction = new()
+    {
+        DisplayName = "action-name-toggle-ninja-gloves",
+        Description = "action-desc-toggle-ninja-gloves",
+        Priority = -13,
+        Event = new ToggleActionEvent()
+    };
+}
+
+/// <summary>
+/// Component for emagging doors on click, when gloves are enabled.
+/// Only works on entities with DoorComponent.
+/// </summary>
+[RegisterComponent]
+public sealed class NinjaDoorjackComponent : Component
+{
+    /// <summary>
+    /// The tag that marks an entity as immune to doorjacking
+    /// </summary>
+    [DataField("emagImmuneTag", customTypeSerializer: typeof(PrototypeIdSerializer<TagPrototype>))]
+    public string EmagImmuneTag = "EmagImmune";
+}
+
+/// <summary>
+/// Component for stunning mobs on click, when gloves are enabled.
+/// Knocks them down for a bit and deals shock damage.
+/// </summary>
+[RegisterComponent]
+public sealed class NinjaStunComponent : Component
+{
+    /// <summary>
+    /// Joules required in the suit to stun someone. Defaults to 10 uses on a small battery.
+    /// </summary>
+    [DataField("stunCharge")]
+    public float StunCharge = 36.0f;
+
+    /// <summary>
+    /// Shock damage dealt when stunning someone
+    /// </summary>
+    [DataField("stunDamage")]
+    public int StunDamage = 5;
+
+    /// <summary>
+    /// Time that someone is stunned for, stacks if done multiple times.
+    /// </summary>
+    [DataField("stunTime")]
+    public TimeSpan StunTime = TimeSpan.FromSeconds(3);
+}
+
+/// <summary>
+/// Component for draining power from APCs/substations/SMESes, when gloves are enabled.
+/// </summary>
+[RegisterComponent]
+public sealed class NinjaDrainComponent : Component
```

Comment:
I think a better implementation would be a generic "BatteryDrainerComponent"(name due to change) that subscribes to both `GotEquipped/UnequippedEvent` and some interaction event. The component would store the container id of the battery and its behavior would be giving the wearer the ability to steal power from a target to the battery id stored in the component and do the same if clicked on something while inhands (maybe behind a datafield?). `ChargerComponent` has a `TransferPower` method that kiiiinda works like this

Reply:
it drains into the ninjas suit battery not an internal battery so making it generic would still be too ninja specific even if it doesnt have ninja in the name

Reply:
Make it have a `ProviderUid` entityuid field, then assign it the hardsuit battery somewhere in ninja code. Then you would use the methods in event handlers like this: `TryUseCharge(component.ProviderUid ??	uid)` (tho doing `component.ProviderUid ??= uid` at the start of the handler might make more sense)

Reply:
also i realise once these features are split out from gloves i can actually add them to the user and wean off interaction relay dependency